app/controllers/canvas_controller.rb

```diff
@@ -114,13 +120,16 @@ def update
 
     # Make sure move parameter is valid
     to_move = {}
-    if can_move_modules(@experiment) && update_params[:move].present?
+    if update_params[:move].present?
       begin
         to_move = JSON.parse(update_params[:move])
         # Okay, JSON parsed!
         unless to_move.is_a?(Hash) &&
                to_move.keys.all? { |k| k.is_a? String } &&
-               to_move.values.all? { |k| k.is_a? String }
+               to_move.values.all? { |k| k.is_a? String } &&
+               to_rename.keys.all? do |id|
+                 can_manage_module?(MyModule.find_by_id(id))
```

Comment:
here you have the action for the module but above in core_api_controller you use a method for experiments can_read_experiment?(task.experiment)
Can you make can_read_module and use it ?

Reply:
If you can read experiment, you can read module. Check the comments on top of can :read_experiment permission.